lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -781,75 +782,144 @@ SILCombiner::visitInjectEnumAddrInst(InjectEnumAddrInst *IEAI) {
   StoreInst *SI = nullptr;
   InitEnumDataAddrInst *DataAddrInst = nullptr;
   ApplyInst *AI = nullptr;
   Operand *EnumInitOperand = nullptr;
-  for (;;) {
-    if (II == IEAI->getParent()->begin())
-      return nullptr;
-    --II;
-    SI = dyn_cast<StoreInst>(&*II);
-    if (SI) {
-      // Find a Store whose destination is taken from an init_enum_data_addr
-      // whose address is same allocation as our inject_enum_addr.
-      DataAddrInst = dyn_cast<InitEnumDataAddrInst>(SI->getDest());
-      if (DataAddrInst && DataAddrInst->getOperand() == IEAI->getOperand())
-        break;
-      SI = nullptr;
+
+  llvm::SmallVector<InitEnumDataAddrInst *, 8> DataAddrVec;
+  llvm::SmallVector<InjectEnumAddrInst *, 8> InjectEnumAddrVec;
+  llvm::SmallSet<SILInstruction *, 32> MayWriteSet;
+  for (auto usersIt : ASO->getUses()) {
+    SILInstruction *currUser = usersIt->getUser();
+    if (currUser->isDeallocatingStack()) {
+      // we don't care about the dealloc stack instructions
+      continue;
     }
-    // Check whether we have an apply initializing the enum.
-    //  %iedai = init_enum_data_addr %enum_addr
-    //         = apply(%iedai,...)
-    //  inject_enum_addr %enum_addr
-    //
-    // We can localize the store to an alloc_stack.
-    // Allowing us to perform the same optimization as for the store.
-    //
-    //  %alloca = alloc_stack
-    //            apply(%alloca,...)
-    //  %load = load %alloca
-    //  %1 = enum $EnumType, $EnumType.case, %load
-    //  store %1 to %nopayload_addr
-    //
-    if ((AI = dyn_cast<ApplyInst>(&*II))) {
-      unsigned ArgIdx = 0;
-      for (auto &Opd : AI->getArgumentOperands()) {
-        // Found an apply that initializes the enum. We can optimize this by
-        // localizing the initialization to an alloc_stack and loading from it.
-        DataAddrInst = dyn_cast<InitEnumDataAddrInst>(Opd.get());
-        if (DataAddrInst && DataAddrInst->getOperand() == IEAI->getOperand() &&
-            ArgIdx < AI->getSubstCalleeType()->getNumIndirectResults()) {
-          EnumInitOperand = &Opd;
-          break;
-        }
-        ++ArgIdx;
-      }
-      // We found an enum initialization.
-      if (EnumInitOperand)
-        break;
-      AI = nullptr;
+    DataAddrInst = dyn_cast<InitEnumDataAddrInst>(currUser);
+    if (DataAddrInst) {
+      DataAddrVec.push_back(DataAddrInst);
+      continue;
     }
+    InjectEnumAddrInst *AddrInst = dyn_cast<InjectEnumAddrInst>(currUser);
+    if (AddrInst) {
+      InjectEnumAddrVec.push_back(AddrInst);
+      continue;
+    }
+    if (currUser->mayWriteToMemory()) {
+      MayWriteSet.insert(currUser);
+    }
+  }
+
+  if (DataAddrVec.size() != 1 || InjectEnumAddrVec.size() != 1) {
+    return nullptr;
   }
-  // Found the store to this enum payload. Check if the store is the only use.
+  assert((InjectEnumAddrVec.front() == IEAI) &&
+         "Found InitEnumDataAddrInst differs from IEAI");
+  DataAddrInst = DataAddrVec.front();
```

Comment:
I would declare DataAddrInst here rather than reuse the one that is declared further up.